grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,280 @@
+        serverHealth = new SettableHealthChecker(false);
+        if (!updateListeners.isEmpty()) {
+            addServerHealthUpdateListener(updateListeners);
+        }
+        this.healthCheckers = ImmutableSet.<ListenableHealthChecker>builder()
+                                          .add(serverHealth)
+                                          .addAll(healthCheckers)
+                                          .build();
+        this.grpcServiceHealthCheckers = grpcServiceHealthCheckers;
+        final HealthCheckUpdateListener healthCheckUpdateListener = provideInternalHealthUpdateListener();
+        setInternalHealthUpdateListener(healthCheckUpdateListener, this.healthCheckers);
+        setInternalHealthUpdateListener(healthCheckUpdateListener, this.grpcServiceHealthCheckers.values());
+    }
+
+    @Override
+    public void check(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
+        final String service = request.getService();
+        final ServingStatus status = checkServingStatus(service);
+        if (status == ServingStatus.SERVICE_UNKNOWN) {
+            responseObserver.onError(Status.NOT_FOUND
+                                             .withDescription(String.format(
+                                                     "The service name(%s) is not registered in this service",
+                                                     service))
+                                             .asRuntimeException());
+            return;
+        }
+        responseObserver.onNext(getHealthCheckResponse(status));
+        responseObserver.onCompleted();
+    }
+
+    @Override
+    public void watch(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
+        final String service = request.getService();
+        final ServingStatus status = checkServingStatus(service);
+        final HealthCheckResponse response = getHealthCheckResponse(status);
+        responseObserver.onNext(response);
+        watchers.put(responseObserver, Maps.immutableEntry(service, response.getStatus()));
+        ((ServerCallStreamObserver<HealthCheckResponse>) responseObserver).setOnCancelHandler(() -> {
+            watchers.remove(responseObserver);
+        });
+    }
+
+    void serviceAdded(ServiceConfig cfg) {
+        if (server != null) {
+            if (server != cfg.server()) {
+                throw new IllegalStateException("Cannot be added to more than one server");
+            } else {
+                return;
+            }
+        }
+
+        server = cfg.server();
+        server.addListener(new ServerListenerAdapter() {
+            @Override
+            public void serverStarted(Server server) {
+                serverHealth.setHealthy(true);
+            }
+
+            @Override
+            public void serverStopping(Server server) {
+                serverHealth.setHealthy(false);
+                watchers.keySet().forEach(StreamObserver::onCompleted);
```

Comment:
Could release the strong references by invoking `watchers.clear()`?

Reply:
Could you add a comment that `NOT_SERVING` is sent to the client by `serverHealth.setHealthy(false);` call for future reviewers? 😉